lib/files/node.ts

```diff
@@ -0,0 +1,166 @@
+/**
+ * @copyright Copyright (c) 2022 John Molakvoæ <skjnldsv@protonmail.com>
+ *
+ * @author John Molakvoæ <skjnldsv@protonmail.com>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+import { basename, extname, dirname } from 'path'
+import { Permission } from '../permissions'
+import NodeData, { Attribute, validateData } from './nodeData'
+
+const davService = 'remote.php/dav'
+
+export enum FileType {
+	Folder = 'folder',
+	File = 'file',
+}
+
+export default abstract class Node {
+	private _data: NodeData
+	private _attributes: Attribute[]
+
+	constructor(data: NodeData) {
+		// Make sure the source includes the current host
+		if (data?.source?.startsWith('/')) {
+			data.source = location.origin + data.source
+		}
+
+		// Validate data
+		validateData(data)
+
+		this._data = data
+		this._attributes = data.attributes || {} as any
+		delete this._data.attributes
+	}
+
+	/**
+	 * Get the source url to this object
+	 */
+	get source(): string {
+		// strip any ending slash
+		return this._data.source.replace(/\/$/i, '')
```

Comment:
does this imply that invalid data was passed to `data.source`?

Reply:
No, it's just to help standardizing the source format a tiny bit. Would you prefer a more restrictive approach and throwing?
I wasn't sure about leaving the ending slash or not. :)